### PR TITLE
Include record errors if there's any on which `destroy!` called

### LINKED ISSUE #33420: `destory!` excludes record errors in exception when exist
### Steps to reproduce
```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :individuals, force: true do |t|
  end

  create_table :addresses, force: true do |t|
    t.integer :individual_id
  end
end

class Individual < ActiveRecord::Base
  has_one :address, dependent: :restrict_with_error
end

class Address < ActiveRecord::Base
  belongs_to :person
end

class BugTest < Minitest::Test
  def test_association_stuff
    individual = Individual.create!
    individual.address = Address.create!

    exception = assert_raise(ActiveRecord::RecordNotDestroyed) do
      individual.destroy!
    end

    assert_equal "Cannot delete record because a dependent address exists", exception.message
  end
end
```
### Expected behavior
"Cannot delete record because a dependent address exists"

### Actual behavior
"Failed to destroy the record"

### System configuration
**Rails version**: 5.x

**Ruby version**: 2.x

### PR DESCRIPTION
**Pull request by al3rez**
### Summary

Fixes #33420 - Check whether the record has errors and include it otherwise
use the default one.

### CODE REVIEW

File: activerecord/lib/active_record/persistence.rb [hunk 1]
```diff
@@ -758,7 +758,8 @@ def verify_readonly_attribute(name)
 
     def _raise_record_not_destroyed
       @_association_destroy_exception ||= nil
-      raise @_association_destroy_exception || RecordNotDestroyed.new("Failed to destroy the record", self)
+      message = self.errors.first.try(:second) || "Failed to destroy the record"
```
Comment (kaspth) [inline activerecord/lib/active_record/persistence.rb]: Not liking `self.errors.first.try(:second)`. `self.` should be removed. `try` should be replaced with `&.`. `first&.second` doesn't seem very clear to me.

Although I haven't looked at this code before, but isn't `@_association_destroy_exception` partly here to facilitate this?

Comment (azbshiri) [reply activerecord/lib/active_record/persistence.rb]: if `@ _association_destroy_exception` can be used, i would just populate that in `destroy!` if errors are present.

Comment (azbshiri) [reply activerecord/lib/active_record/persistence.rb]: I also checked codebase to find `try()` usage and be consistent but if i could use `&.` it'd be great.

**Comment by schneems** [comment]
Still working on this one?

**Comment by azbshiri** [comment]
I'd be happy to if you give me some hints.

File: activerecord/lib/active_record/persistence.rb (hunk 1 above)
Comment (azbshiri) [reply activerecord/lib/active_record/persistence.rb]: Where should I populate `@ _association_destroy_exception`?

https://github.com/rails/rails/blob/2a470d73a75ebf8cd7975e469bd82586d9234442/activerecord/lib/active_record/callbacks.rb#L306

**Comment by oniofchaos** [comment]
> self. should be removed. try should be replaced with &.

I think at least making these two changes will make the code more readable and hopefully more hints  will reveal themselves from that point

**Comment by azbshiri** [comment]
@schneems @kaspth Refactored

**Comment by azbshiri** [comment]
@kamipo I got some failures due to connection adapters which is irrelevant to my changes? How can I restart Travis?

**Pull request closed without merging**